// discord.js-anticrash@4.0.3-devbuild/index.js
const {
  WebhookClient,
  EmbedBuilder,
  AttachmentBuilder,
} = require("discord.js");
const colors = require("colors");

const DISCORD_API_ERRORS = {
  unknown: (code) =>
    !(
      (code >= 10001 && code <= 10418) ||
      (code >= 4000 && code <= 4014) ||
      code === 429 ||
      code === 4001
    ),
  invalidData: (code) => code >= 10001 && code <= 10418,
  permission: (code) => code >= 50001 && code <= 50015,
  rateLimit: (code) => code === 429,
  gatewayErrors: (code) => code >= 4000 && code <= 4014,
  invalidToken: (code) => code === 4001,
};

const DEFAULT_CONFIG = {
  embedColor: "#ff0000",
  embedTitle: "Error",
  webhookUsername: "Error",
  embedAvatarUrl: "",
  errorLog: true,
  warningLog: true,
  infoLog: true,
  webhookUrl: process.env.WEBHOOK_URL || "",
};

const timestamp = () => new Date().toLocaleString();
const formatError = (error) =>
  error instanceof Error
    ? `${error.message} (${
        error.stack.split("\n")[1]?.trim() || "No stack trace"
      })`
    : error.toString() || "No error message provided.";

const log = {
  error: logError,
  warning: logWarning,
  info: logInfo,
};

function generateBorder(borderStyle, borderLength) {
  const { top, horizontal, vertical, bottom } = borderStyle;
  const borderLine = horizontal.repeat(borderLength);
  const topBorder = top + borderLine + top;
  const bottomBorder = bottom + borderLine + bottom;
  return { topBorder, bottomBorder, vertical };
}

function logError(error, details = "") {
  const timestamp = new Date().toLocaleString();
  const errorMessage =
    error instanceof Error
      ? error.message
      : error.toString() || "No error message provided.";
  const origin =
    error instanceof Error ? `(${error.stack.split("\n")[1].trim()})` : "";
  const borderLength = Math.max(
    errorMessage.length + 20,
    timestamp.length + origin.length + 20
  );
  const paddingLength = Math.max(
    0,
    borderLength - errorMessage.length - timestamp.length - 10
  );

  const borderStyle = {
    top: "╔",
    horizontal: "═",
    vertical: "║",
    bottom: "╚",
  };

  const { topBorder, bottomBorder, vertical } = generateBorder(
    borderStyle,
    borderLength
  );
  const paddedError =
    " ".repeat(2) + colors.gray(errorMessage) + " ".repeat(paddingLength);
  const paddedOrigin = colors.red(origin) + " ".repeat(paddingLength);

  console.error(topBorder);
  console.error(
    vertical + colors.bgRed.bold(` ERROR `) + colors.red(` [${timestamp}] `)
  );
  console.error(vertical + colors.red.bold(paddedError) + vertical);
  console.error(vertical + colors.red.bold(paddedOrigin) + vertical);
  if (details) {
    console.error(
      vertical +
        " ".repeat(2) +
        colors.red.bold(details) +
        " ".repeat(paddingLength) +
        vertical
    );
  }
  console.error(bottomBorder);
}

function logWarning(warning, details = "") {
  const timestamp = new Date().toLocaleString();
  const borderLength = warning.length + 30;
  const paddingLength = Math.max(
    0,
    borderLength - warning.length - timestamp.length - 11
  );

  const borderStyle = {
    top: "╔",
    horizontal: "═",
    vertical: "║",
    bottom: "╚",
  };

  const { topBorder, bottomBorder, vertical } = generateBorder(
    borderStyle,
    borderLength
  );
  const paddedWarning = " ".repeat(2) + warning + " ".repeat(paddingLength);
  const timestampLine = " ".repeat(2) + timestamp + " ".repeat(2);

  console.warn(topBorder);
  console.warn(
    vertical +
      colors.bgYellow.black.bold(` WARNING `) +
      colors.yellow(` [${timestamp}] `)
  );
  console.warn(vertical + colors.yellow.bold(paddedWarning) + vertical);
  if (details) {
    console.warn(
      vertical +
        " ".repeat(2) +
        colors.yellow.bold(details) +
        " ".repeat(paddingLength) +
        vertical
    );
  }
  console.warn(vertical + colors.yellow.bold(timestampLine) + vertical);
  console.warn(bottomBorder);
}

function logInfo(info) {
  const timestamp = new Date().toLocaleString();
  const borderLength = info.length + 22;
  const paddingLength = Math.max(
    0,
    borderLength - info.length - timestamp.length - 11
  );

  const borderStyle = {
    top: "╔",
    horizontal: "═",
    vertical: "║",
    bottom: "╚",
  };

  const { topBorder, bottomBorder, vertical } = generateBorder(
    borderStyle,
    borderLength
  );
  const paddedInfo = " ".repeat(2) + info + " ".repeat(paddingLength);
  const timestampLine = " ".repeat(2) + timestamp + " ".repeat(2);

  console.log(topBorder);
  console.log(
    vertical + colors.bgCyan.bold(` INFO `) + colors.cyan(` [${timestamp}] `)
  );
  console.log(vertical + colors.cyan.bold(paddedInfo) + vertical);
  console.log(vertical + colors.cyan.bold(timestampLine) + vertical);
  console.log(bottomBorder);
}

const categorizeDiscordError = (error) => {
  if (error.name === "DiscordAPIError") {
    for (const [key, checkFunction] of Object.entries(DISCORD_API_ERRORS)) {
      if (checkFunction(error.code)) return key;
    }
    return "Unknown Discord API Error";
  }
  return error.name.replace(/Error$/, "") || "General Discord.js Error";
};

const generatePossibleSolutions = (errorCategory) => {
  const solutions = {
    "Rate Limit Error": [
      "1. Implement rate limit handling (e.g., delays, queues).",
      "2. Reduce API requests.",
      "3. Cache data to avoid unnecessary requests.",
    ],
    "Bot Permissions Error": [
      "1. Grant the bot the required permissions in the server settings.",
      "2. Check the order of roles if permissions are conflicting.",
    ],
    "Sharding Error": [
      "1. Verify sharding configuration.",
      "2. Ensure consistent code across shards.",
      "3. Debug communication between shards.",
    ],
    "Voice Connection Error": [
      "1. Check network connectivity.",
      "2. Verify voice channel permissions.",
      "3. Update discord.js and/or voice dependencies.",
    ],
    "HTTP Request Error": [
      "1. Check network connectivity.",
      "2. Verify API endpoint URL.",
      "3. Debug HTTP requests and responses.",
    ],
    "Authentication Error": [
      "1. Verify bot token validity.",
      "2. Check bot permissions and scopes.",
      "3. Ensure correct authentication methods.",
    ],
    "Missing Intents Error": [
      "1. Ensure the bot has the necessary intents declared in its configuration.",
      "2. Update the `intents` array in your `client.options` object to include the missing intents.",
      "3. Restart your bot after making changes to the intents.",
    ],
    "General Discord.js Error": [
      "1. Consult Discord.js documentation for potential issues with the specific API method or event you are using.",
      "2. Verify the Discord API response and its structure.",
      "3. Check community resources and support channels for similar issues.",
    ],
    "Unknown Discord API Error": [
      "1. Review error message and stack trace for details.",
      "2. Check documentation for potential issues.",
      "3. Consult community resources or support channels.",
    ],
  };

  return solutions[errorCategory] || solutions["General Discord.js Error"];
};

const sendErrorMessage = async (error, eventType, additionalInfo = "", config = {}) => {
  try {
    const embedConfig = { ...DEFAULT_CONFIG, ...config };
    if (!embedConfig.webhookUrl) {
      throw new Error("Webhook URL is required.");
    }

    const webhook = new WebhookClient({ url: embedConfig.webhookUrl });
    const errorMessage = formatError(error);

    const embed = new EmbedBuilder()
      .setColor(embedConfig.embedColor)
      .setTimestamp()
      .setTitle(`${eventType.toUpperCase()} - ${embedConfig.embedTitle}`)
      .setAuthor({ name: "Discord.js ‣ Anticrash" })
      .setFooter({ text: "Powered by discord.js-anticrash" })
      .addFields([
        { name: "__Event Type__", value: `\`${eventType}\``, inline: true },
        {
          name: "__Message__",
          value: `**\`${additionalInfo}\`**`,
          inline: true,
        },
        {
          name: "__Error Type__",
          value: `\`\`\`\n${error.constructor.name}\n\`\`\``,
          inline: true,
        },
        {
          name: "__Error Category__",
          value: `\`${categorizeDiscordError(error)}\``,
          inline: true,
        },
        {
          name: "__Possible Solutions__",
          value: generatePossibleSolutions(categorizeDiscordError(error)).join(
            "\n"
          ),
          inline: false,
        },
      ])
      .setDescription(
        `__Detailed__\n${errorMessage
          .split("\n")
          .slice(0, 5)
          .map((line, index) => `\`\`\`\n${index + 1}. ${line.trim()}\n\`\`\``)
          .join("\n")}`
      );

    await webhook.send({
      username: embedConfig.webhookUsername,
      avatarURL: embedConfig.embedAvatarUrl,
      embeds: [embed],
    });

    if (errorMessage.length > 4076) {
      await webhook.send({
        username: embedConfig.webhookUsername,
        avatarURL: embedConfig.embedAvatarUrl,
        files: [
          new AttachmentBuilder()
            .setName("error.txt")
            .setFile(Buffer.from(errorMessage, "utf-8"))
            .setSpoiler(false),
        ],
      });
    }
  } catch (error) {
    log.error(`Error sending error message: ${error.message}`);
  }
};


const errorHandling = async (client, config = {}) => {
  const embedConfig = { ...DEFAULT_CONFIG, ...config };

  if (!embedConfig.webhookUrl) {
    log.error("Webhook URL is required.");
    throw new Error("Webhook URL is required.");
  }

  const webhook = new WebhookClient({ url: embedConfig.webhookUrl });

  const processEventListeners = {
    unhandledRejection: {
      level: "error",
      listener: async (reason, promise) => {
        await sendErrorMessage(
          reason,
          "Unhandled Rejection",
          `Promise: ${promise}`,
          embedConfig
        );
        log.error(`Unhandled Rejection: ${formatError(reason)}`);
      },
    },
    uncaughtException: {
      level: "error",
      listener: async (error, origin) => {
        await sendErrorMessage(
          error,
          "Uncaught Exception",
          `Origin: ${origin}`,
          embedConfig
        );
        log.error(`Uncaught Exception: ${formatError(error)}`);
      },
    },
    uncaughtExceptionMonitor: {
      level: "warning",
      listener: async (error, origin) => {
        await sendErrorMessage(
          error,
          "Uncaught Exception Monitor",
          `Origin: ${origin}`,
          embedConfig
        );
        log.warning(`Uncaught Exception Monitor: ${formatError(error)}`);
      },
    },
    warning: {
      level: "warning",
      listener: async (warning) => {
        await sendErrorMessage(warning, "Warning", "", embedConfig);
        log.warning(`Warning: ${warning}`);
      },
    },
    exit: {
      level: "info",
      listener: async (code, signal) => {
        const message = `Exiting with code ${code} and signal ${signal}`;
        await sendErrorMessage(new Error(message), "Exit", message, embedConfig);
        log.info(message);
      },
    },
  };

  for (const [event, listenerConfig] of Object.entries(processEventListeners)) {
    process.on(event, async (...args) => {
      try {
        await listenerConfig.listener(...args);
      } catch (error) {
        log.error(
          `Error in process event listener: ${event} - ${formatError(error)}`
        );
      }
    });
  }
};

module.exports = { errorHandling };
